include/api.php

```diff
@@ -3276,72 +3276,706 @@ function api_oauth_access_token($type) {
+		$acl_input_error |= CheckACLInput($allow_gid);
+		$acl_input_error |= CheckACLInput($deny_gid);
+		if ($acl_input_error)
+			throw new BadRequestException("acl data invalid");
+
+		// now let's upload the new media in create-mode
+		if ($mode == "create") {
+			$media = $_FILES['media'];
+			$data = SaveMediaToDatabase("photo", $media, $type, $album, trim($allow_cid), trim($deny_cid), trim($allow_gid), trim($deny_gid), $desc, $visibility);
+
+			// return success of updating or error message
+			if (!is_null($data)) {
+				return api_format_data("photo_create", $type, $data);
+			}
+			else {
+				throw new BadRequestException("unknown error");
+			}
+		}
+
+		// now let's do the changes in update-mode
+		if ($mode == "update") {
+			$sql_extra = "";
+
+			if (!is_null($desc)) {
+				$sql_extra .= (($sql_extra != "") ? " ," : "") . "`desc` = '$desc'";
+			}
+
+			if (!is_null($album_new)) {
+				$sql_extra .= (($sql_extra != "") ? " ," : "") . "`album` = '$album_new'";
+			}
+
+			if (!is_null($allow_cid)) {
+				$allow_cid = trim($allow_cid);
+				$sql_extra .= (($sql_extra != "") ? " ," : "") . "`allow_cid` = '$allow_cid'";
+			}
+
+			if (!is_null($deny_cid)) {
+				$deny_cid = trim($deny_cid);
+				$sql_extra .= (($sql_extra != "") ? " ," : "") . "`deny_cid` = '$deny_cid'";
+			}
+
+			if (!is_null($allow_gid)) {
+				$allow_gid = trim($allow_gid);
+				$sql_extra .= (($sql_extra != "") ? " ," : "") . "`allow_gid` = '$allow_gid'";
+			}
+
+			if (!is_null($deny_gid)) {
+				$deny_gid = trim($deny_gid);
+				$sql_extra .= (($sql_extra != "") ? " ," : "") . "`deny_gid` = '$deny_gid'";
+			}
+
+			$result = false;
+			if ($sql_extra != "") {
+				$nothingtodo = false;
+				$result = q("UPDATE `photo` SET %s, `edited`='%s' WHERE `uid` = %d AND `resource-id` = '%s' AND `album` = '%s'",
+						$sql_extra,
+						datetime_convert(),   // update edited timestamp
+						intval(api_user()),
+						dbesc($photo_id),
+						dbesc($album));
+			} else {
+				$nothingtodo = true;
+			}
+
+			if(x($_FILES,'media')) {
+				$nothingtodo = false;
+				$media = $_FILES['media'];
+				$data = SaveMediaToDatabase("photo", $media, $type, $album, $allow_cid, $deny_cid, $allow_gid, $deny_gid, $desc, 0, $visibility, $photo_id);
+				if (!is_null($data))
+					return api_format_data("photo_update", $type, $data);
+			}
+
+			// return success of updating or error message
+			if ($result) {
+				$answer = array('result' => 'updated', 'message' => 'Image id `' . $photo_id . '` has been updated.');
+				return api_format_data("photo_update", $type, array('$result' => $answer));
+			}
+			else {
+				if ($nothingtodo) {
+					$answer = array('result' => 'cancelled', 'message' => 'Nothing to update for image id `' . $photo_id . '`.');
+					return api_format_data("photo_update", $type, array('$result' => $answer));
+				}
+				throw new BadRequestException("unknown error");
+			}
+		}
+		throw new BadRequestException("unknown error");
```

Comment:
this is again just to avoid a warning that not all code paths return a value. This should never be reached, output changed to "unknown error - this error on uploading or updating a photo should never happen"